### ISSUE TITLE
Syntax error due to parenthesis.

### ISSUE THREAD

**Issue by VincentCheval** [FStarLang/FStar#59]
Hello,

I think I found a small bug into the parser. The following code raises a syntax error.
```
module Helios

val length : list 'a -> Tot nat
let rec length l = match l with
  | [] -> 0
  | hd::tl -> 1 + length tl  

val lem_len_nil: unit -> Lemma (length [] = 0)
let lem_len_nil () = () 

val lem_len_cons: hd:'a -> (tl:list 'a) -> Lemma (length (hd::tl) = (length tl) + 1)
let lem_len_cons hd tl = ()
```

When I replace the line 
```
val lem_len_cons: (hd:'a) -> (tl:list 'a) -> Lemma (length (hd::tl) = (length tl) + 1)
```
by 
```
val lem_len_cons: hd:'a -> tl:list 'a -> Lemma (length (hd::tl) = (length tl) + 1)
```
it works well. 

Best regards,
Vincent Cheval

**Comment by catalin-hritcu** [comment]
I agree with Nik, it's commonplace that parentheses delimit binding scopes. For instance `\x. x x` is different than `(\x. x) x`.

**Comment by nikswamy** [comment]
I might be backtracking on this one ... 
This discrepancy between the syntax of binders in type and let definitions and the binders in arrow types is annoying.

**Comment by nikswamy** [comment]
Not sure about how best to resolve this. We have some conflicting requirements. 

1. We like the `x:t -> t'` syntax, particularly when it is used as `x:t{phi} -> t'`.

2. As such, this type is currently accepted: `a:Type -> (x:a -> int) -> int`, where the second argument is an unnamed function, with a named argument. 

3. But, this syntax is rejected: `let f (a:Type) (x:a -> int) = 0` ... the second argument is an unnamed function ... and in a let binding, all the parameters have to be named. For the record, the same comment applies to `type f (a:Type) (x:a -> int) = int`


How can we resolve this? Suggestions most welcome.

**Comment by nikswamy** [comment]
argh. this is screwed up ... please hold off on comments until I fix my last comment

**Comment by nikswamy** [comment]
Let me try again (working through this with Jonathan) (It's a bad sign that I am so confused about the syntax myself : )

This parses:
```
let f (x:int) (y:int -> int) = y
```

But, this doesn't:
```
type g = (x:int) -> (y:int -> int) -> int
```

The failure is in the parsing of `(x:int)`: The grammar attempts to parse this as an atomic type (delimited by the parens), then expects to parse a refinement type because of the binder `x:int` and then fails because there is no following refinement. The variant below parses:

```
type h = (x:int{x > 0}) -> (y:int -> int) -> int
```

As does this variant:
```
type g' = x:int -> (y:int -> int) -> int
```

However, it parses with different scoping rules. So, this version parses but fails to type-check (actually, fails during name-binding while desugaring). Specifically, the `y` names only the argument `int` not the entire function.

```
assume type vector : int -> Type
type g'' = x:int -> (y:int -> int) -> vector (y 0)
```

This is consistent with `g'` and `h` ... but different than `f`. Perhaps that's ok ... perhaps not.

**Comment by nikswamy** [comment]
Again, with Jonathan:

FYI: A further inconsistency was that the parsing of the parameters of `let`-bindings and the parameters in `type` definitions was different. This did not parse until my most recent commit 068990c5d8581effb7a51cd200c4f146319eeaba

```
type ff (x:int) (y:int -> Tot int) = vector (y 0)
```

However, it does parse and type-check now. 

So does this:
```
type ff (x:int) (y:(int -> Tot int)) = vector (y 0)
```

**Comment by nikswamy** [comment]
So, what are our options:

1. We live with the difference between the syntax of arrows and of let/type definitions. This could be mitigated with some tricks (see below).

2. We insist on parenthesis in the binders even in arrow types. So, you would have to write `(x:int{x > 0}) -> vector x`. This would also require a change in binding rules for arrow types, where even though it is delimited by parens, the scope of `x` would span over both the refinement and the RHS of the arrow. The upside is that let/type becomes consistent with arrow. The downside is that we need to write more parens in arrow types ... which is burdensome.

**Comment by catalin-hritcu** [comment]
I don't have any opinion one way or the other, but let me try to cheer you up: https://xkcd.com/297/

**Comment by nikswamy** [comment]
Resolving this mess in a465a241a90e064faa347f9deacde8eec12a1c70

**Comment by nikswamy** [comment]
Here's what the parser accepts now:

```
module Test
assume type tint : int -> Type

type t1 = x:int{x >= 0} -> tint x
type t2 = x:int -> res:tint x
type t3 = (a:Type) -> (x:int) -> result:list a{result=[] /\ x=18}
type t4 = x:nat & y:nat{y > x}
type t4' = (x:nat) & y:nat{y > x}
type t4'' = (x:nat) & (y:nat{y > x}) //parentheses are not significant
type t5 = x:int{x=0} * y:int
//type t5' = x:int{x=0} * tint x
//parses but fails name-binding; x is not in scope in the 2nd component
type t6 = (int * int) -> int
type t7 = int * int -> int

type t8 = x:int * int -> int
//type t8' = x:int * int -> tint x //x is not in scope to the right of the arrow
type t8' = x:(int * int) -> tint (fst x)

type t9 = x:int & tint x -> int
type t9' = y:(x:int & tint x) -> tint (dfst y)

type ta = (x:int) -> (y:int -> tint y) -> tint x
type tb = x:int -> (y:int -> tint y) -> tint x
type tc = x:int -> f:(y:int -> Tot int) -> tint (f 0)

let f1 (x: y:int &  z:tint y{z=z /\ y=0}) = x
let f2 (x: y:int *  z:tint 0{z=z}) = x
let f3 (x: y:int -> z:tint y{z=z /\ y=0}) = x
let f4 (#a:Type) (l:list a) (#x:int) (v:tint x) = l
let f5 (l:list int) (v:tint 0) = f4 l v

type s0 (x:int) (y:int -> Tot int) = tint (y 0)
type s1 (x: y:int &  z:tint y{z=z /\ y=0}) = tint (dfst x)
type s2 (x: y:int *  z:tint 0{z=z}) = tint (fst x)
type s3 (x: y:int -> z:tint y{z=z /\ y=0}) = tint ((fun x -> 0) x)
```

**Comment by nikswamy** [comment]
The main takeaway point is that the discrepancy between the syntax of arrows and type/let-bindings has been resolved. 

There is, however, a difference in the scoping rules for those constructs, particular at higher order. 

Specifically:

In the arrow type below, `x` names the formal parameter of the unnamed function-typed argument. It is in scope in `t1` but *not* in `t2`.
```
 (x:t0 -> t1) -> t2
```
As an aside, the same rule applies to the &` constructor, for dependent tuples.

In a let-binding, type-definition, or quantifier, each argument carries a mandatory name. So, the only way to parse

```
type (x:t0 -> t1) = t2
```
is for `x` to name the function `t0 -> t1` and for `x` to be in scope in `t2`.

For first-order types, this distinction is irrelevant. For example, in both cases below, `x` names the argument `t0` and is in scope in `t1`

```
(x:t0) -> t1
type t (x:t0) = t1
```
